File: frontend/src/pages/FAQ.js
import React, { useState, useEffect } from 'react';
import { useQuery } from 'react-query';
import { 
  FiSearch, 
  FiHelpCircle, 
  FiChevronDown, 
  FiChevronUp,
  FiExternalLink,
  FiMessageSquare,
  FiBookOpen
} from 'react-icons/fi';
import { faqAPI } from '../services/api';
import toast from 'react-hot-toast';

const FAQ = () => {
  const [searchQuery, setSearchQuery] = useState('');
  const [selectedCategory, setSelectedCategory] = useState('');
  const [expandedItems, setExpandedItems] = useState(new Set());
  const [suggestedQuestions, setSuggestedQuestions] = useState([]);

  // Fetch FAQ categories
  const { data: categories } = useQuery(
    'faq-categories',
    () => faqAPI.getCategories(),
    { retry: 1 }
  );

  // Fetch FAQ search results
  const { data: searchResults, isLoading: searchLoading } = useQuery(
    ['faq-search', searchQuery, selectedCategory],
    () => faqAPI.search(searchQuery, selectedCategory, 10),
    { 
      retry: 1,
      enabled: searchQuery.length > 0
    }
  );

  // Fetch suggested questions
  const { data: suggestions } = useQuery(
    'faq-suggestions',
    () => faqAPI.getSuggestions(''),
    { retry: 1 }
  );

  // Fetch all FAQs if no search
  const { data: allFAQs, isLoading: allFAQsLoading } = useQuery(
    'faq-all',
    () => faqAPI.getAll(),
    { 
      retry: 1,
      enabled: searchQuery.length === 0
    }
  );

  useEffect(() => {
    if (suggestions?.suggestions) {
      setSuggestedQuestions(suggestions.suggestions);
    }
  }, [suggestions]);

  const toggleExpanded = (itemId) => {
    const newExpanded = new Set(expandedItems);
    if (newExpanded.has(itemId)) {
      newExpanded.delete(itemId);
    } else {
      newExpanded.add(itemId);
    }
    setExpandedItems(newExpanded);
  };

  const handleSearch = (e) => {
    e.preventDefault();
    if (searchQuery.trim()) {
      // Search is handled by the query
      toast.success(`Searching for: ${searchQuery}`);
    }
  };

  const handleSuggestionClick = (suggestion) => {
    setSearchQuery(suggestion);
    toast.success(`Searching for: ${suggestion}`);
  };

  const faqItems = searchQuery.length > 0 ? searchResults?.results : allFAQs?.faqs || [];

  const quickStartGuide = [
    {
      step: 1,
      title: 'Upload Your Documents',
      description: 'Upload PDF, Excel, or CSV files containing your financial data',
      icon: '📄'
    },
    {
      step: 2,
      title: 'Ask Questions',
      description: 'Use natural language to ask questions about your financial data',
      icon: '💬'
    },
    {
      step: 3,
      title: 'Get Insights',
      description: 'Receive detailed analysis, ratios, and forecasts',
      icon: '📊'
    },
    {
      step: 4,
      title: 'Use Voice Assistant',
      description: 'Speak your questions using the voice assistant feature',
      icon: '🎤'
    }
  ];

  const commonQuestions = [
    'What is FinMDA-Bot?',
    'How do I upload documents?',
    'What financial ratios can you calculate?',
    'Can you predict future performance?',
    'How do I use the voice assistant?',
    'Is my data secure?',
    'What file formats are supported?',
    'How accurate are the financial calculations?'
  ];

  return (
    <div className="space-y-6">
      {/* Header */}
      <div className="text-center">
        <h1 className="text-3xl font-bold text-gray-900">Help & FAQ</h1>
        <p className="text-gray-600 mt-2">Find answers to common questions about FinMDA-Bot</p>
      </div>

      {/* Search Bar */}
      <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6">
        <form onSubmit={handleSearch} className="flex items-center space-x-4">
          <div className="flex-1 relative">
            <FiSearch className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400" size={20} />
            <input
              type="text"
              value={searchQuery}
              onChange={(e) => setSearchQuery(e.target.value)}
              placeholder="Search for help topics..."
              className="w-full pl-10 pr-4 py-3 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-primary-500 focus:border-transparent"
            />
          </div>
          <select
            value={selectedCategory}
            onChange={(e) => setSelectedCategory(e.target.value)}
            className="px-4 py-3 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-primary-500"
          >
            <option value="">All Categories</option>
            {categories?.categories?.map((category) => (
              <option key={category.value} value={category.value}>
                {category.name}
              </option>
            ))}
          </select>
          <button
            type="submit"
            className="px-6 py-3 bg-primary-500 text-white rounded-lg hover:bg-primary-600 transition-colors duration-200"
          >
            Search
          </button>
        </form>
      </div>

      {/* Quick Start Guide */}
      <div className="bg-gradient-to-r from-primary-500 to-primary-600 rounded-lg p-6 text-white">
        <div className="flex items-center space-x-3 mb-4">
          <FiBookOpen size={24} />
          <h2 className="text-xl font-semibold">Quick Start Guide</h2>
        </div>
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4">
          {quickStartGuide.map((step) => (
            <div key={step.step} className="bg-white bg-opacity-20 rounded-lg p-4">
              <div className="flex items-center space-x-3 mb-2">
                <span className="text-2xl">{step.icon}</span>
                <span className="font-semibold">Step {step.step}</span>
              </div>
              <h3 className="font-medium mb-1">{step.title}</h3>
              <p className="text-sm text-primary-100">{step.description}</p>
            </div>
          ))}
        </div>
      </div>

      {/* Suggested Questions */}
      {suggestedQuestions.length > 0 && (
        <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-6">
          <h2 className="text-lg font-semibold text-gray-900 mb-4">Popular Questions</h2>
          <div className="grid grid-cols-1 md:grid-cols-2 gap-3">
            {suggestedQuestions.slice(0, 8).map((question, index) => (
              <button
                key={index}
                onClick={() => handleSuggestionClick(question)}
                className="text-left p-3 rounded-lg border border-gray-200 hover:border-primary-300 hover:bg-primary-50 transition-colors duration-200"
              >
                <p className="text-sm font-medium text-gray-900">{question}</p>
              </button>
            ))}
          </div>
        </div>
      )}

      {/* FAQ Results */}
      <div className="bg-white rounded-lg shadow-sm border border-gray-200">
        <div className="px-6 py-4 border-b border-gray-200">
          <h2 className="text-lg font-semibold text-gray-900">
            {searchQuery ? `Search Results for "${searchQuery}"` : 'Frequently Asked Questions'}
          </h2>
          {searchQuery && (
            <p className="text-sm text-gray-600 mt-1">
              {searchResults?.count || 0} results found
            </p>
          )}
        </div>

        {searchLoading || allFAQsLoading ? (
          <div className="flex items-center justify-center py-12">
            <div className="animate-spin rounded-full h-8 w-8 border-b-2 border-primary-500"></div>
          </div>
        ) : faqItems.length === 0 ? (
          <div className="text-center py-12">
            <FiHelpCircle size={48} className="text-gray-400 mx-auto mb-4" />
            <h3 className="text-lg font-medium text-gray-900 mb-2">No Results Found</h3>
            <p className="text-gray-600 mb-4">
              {searchQuery ? 'Try different search terms or browse all FAQs' : 'No FAQs available'}
            </p>
            {searchQuery && (
              <button
                onClick={() => setSearchQuery('')}
                className="text-primary-600 hover:text-primary-700 font-medium"
              >
                Clear search and view all FAQs
              </button>
            )}
          </div>
        ) : (
          <div className="divide-y divide-gray-200">
            {faqItems.map((faq) => (
              <div key={faq.id} className="p-6">
                <button
                  onClick={() => toggleExpanded(faq.id)}
                  className="w-full flex items-center justify-between text-left"
                >
                  <div className="flex-1">
                    <h3 className="text-lg font-medium text-gray-900 mb-1">
                      {faq.question}
                    </h3>
                    <div className="flex items-center space-x-4 text-sm text-gray-500">
                      <span className="capitalize">{faq.category}</span>
                      <span className="capitalize">{faq.difficulty}</span>
                      {faq.score && (
                        <span>Relevance: {(faq.score * 100).toFixed(0)}%</span>
                      )}
                    </div>
                  </div>
                  <div className="ml-4">
                    {expandedItems.has(faq.id) ? (
                      <FiChevronUp size={20} className="text-gray-400" />
                    ) : (
                      <FiChevronDown size={20} className="text-gray-400" />
                    )}
                  </div>
                </button>

                {expandedItems.has(faq.id) && (
                  <div className="mt-4 pl-0">
                    <div className="prose prose-sm max-w-none">
                      <p className="text-gray-700 leading-relaxed">{faq.answer}</p>
                    </div>
                    
                    {faq.examples && faq.examples.length > 0 && (
                      <div className="mt-4">
                        <h4 className="text-sm font-medium text-gray-900 mb-2">Examples:</h4>
                        <ul className="list-disc list-inside text-sm text-gray-600 space-y-1">
                          {faq.examples.map((example, index) => (
                            <li key={index}>{example}</li>
                          ))}
                        </ul>
                      </div>
                    )}

                    <div className="mt-4 flex items-center space-x-4">
                      <button className="flex items-center space-x-2 text-primary-600 hover:text-primary-700 text-sm font-medium">
                        <FiMessageSquare size={16} />
                        <span>Ask a follow-up question</span>
                      </button>
                      <button className="flex items-center space-x-2 text-gray-600 hover:text-gray-700 text-sm font-medium">
                        <FiExternalLink size={16} />
                        <span>Learn more</span>
                      </button>
                    </div>
                  </div>
                )}
              </div>
            ))}
          </div>
        )}
      </div>

      {/* Contact Support */}
      <div className="bg-gray-50 rounded-lg p-6">
        <div className="text-center">
          <h3 className="text-lg font-semibold text-gray-900 mb-2">Still need help?</h3>
          <p className="text-gray-600 mb-4">
            Can't find what you're looking for? Our support team is here to help.
          </p>
          <div className="flex justify-center space-x-4">
            <button className="px-6 py-2 bg-primary-500 text-white rounded-lg hover:bg-primary-600 transition-colors duration-200">
              Contact Support
            </button>
            <button className="px-6 py-2 bg-white border border-gray-300 text-gray-700 rounded-lg hover:bg-gray-50 transition-colors duration-200">
              Submit Feedback
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default FAQ;

File: backend/app/api/endpoints/faq.py
"""
FAQ API endpoints.
"""
from fastapi import APIRouter, HTTPException, Query
from typing import Dict, List, Any, Optional
import logging

from app.services.faq_service import FAQService, FAQCategory
<<<<<<< HEAD
=======
from app.schemas import ChatRequest, ChatResponse
>>>>>>> 5c3a0a0f3539fc0d352cb6c8a94fa282129f33e9

router = APIRouter()
logger = logging.getLogger(__name__)

# Initialize FAQ service
faq_service = FAQService()


@router.get("/search")
async def search_faqs(
    query: str = Query(..., description="Search query"),
    category: Optional[str] = Query(None, description="FAQ category filter"),
    limit: int = Query(5, description="Maximum number of results")
) -> Dict[str, Any]:
    """Search FAQs based on query."""
    try:
        # Convert category string to enum if provided
        faq_category = None
        if category:
            try:
                faq_category = FAQCategory(category)
            except ValueError:
                raise HTTPException(status_code=400, detail=f"Invalid category: {category}")
        
        # Search FAQs
        results = faq_service.search_faqs(
            query=query,
            category=faq_category,
            limit=limit
        )
        
        return {
            'query': query,
            'category': category,
            'results': results,
            'count': len(results)
        }
        
    except Exception as e:
        logger.error(f"Error searching FAQs: {str(e)}")
        raise HTTPException(status_code=500, detail=f"FAQ search failed: {str(e)}")


@router.get("/{faq_id}")
async def get_faq(faq_id: str) -> Dict[str, Any]:
    """Get FAQ by ID."""
    try:
        faq = faq_service.get_faq_by_id(faq_id)
        if not faq:
            raise HTTPException(status_code=404, detail=f"FAQ not found: {faq_id}")
        
        return faq
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting FAQ {faq_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get FAQ: {str(e)}")


@router.get("/category/{category}")
async def get_faqs_by_category(category: str) -> Dict[str, Any]:
    """Get FAQs by category."""
    try:
        # Convert category string to enum
        try:
            faq_category = FAQCategory(category)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid category: {category}")
        
        # Get FAQs by category
        results = faq_service.get_faqs_by_category(faq_category)
        
        return {
            'category': category,
            'faqs': results,
            'count': len(results)
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting FAQs by category {category}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get FAQs by category: {str(e)}")


@router.get("/")
async def get_all_faqs() -> Dict[str, Any]:
    """Get all FAQs."""
    try:
        all_faqs = []
        for faq in faq_service.faq_items.values():
            all_faqs.append({
                'id': faq.id,
                'question': faq.question,
                'answer': faq.answer,
                'category': faq.category.value,
                'difficulty': faq.difficulty
            })
        
        return {
            'faqs': all_faqs,
            'count': len(all_faqs)
        }
        
    except Exception as e:
        logger.error(f"Error getting all FAQs: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get all FAQs: {str(e)}")


@router.get("/categories/list")
async def get_categories() -> Dict[str, Any]:
    """Get all FAQ categories."""
    try:
        categories = faq_service.get_all_categories()
        return {
            'categories': categories,
            'count': len(categories)
        }
        
    except Exception as e:
        logger.error(f"Error getting categories: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get categories: {str(e)}")


@router.get("/suggestions/questions")
async def get_suggested_questions(
    context: str = Query("", description="Context for suggestions")
) -> Dict[str, Any]:
    """Get suggested questions based on context."""
    try:
        suggestions = faq_service.get_suggested_questions(context)
        return {
            'context': context,
            'suggestions': suggestions,
            'count': len(suggestions)
        }
        
    except Exception as e:
        logger.error(f"Error getting suggested questions: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get suggested questions: {str(e)}")


@router.post("/add")
async def add_faq(
    question: str,
    answer: str,
    category: str,
    keywords: List[str],
    related_questions: Optional[List[str]] = None,
    examples: Optional[List[str]] = None,
    difficulty: str = "beginner"
) -> Dict[str, Any]:
    """Add a new FAQ item."""
    try:
        # Convert category string to enum
        try:
            faq_category = FAQCategory(category)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid category: {category}")
        
        # Add FAQ
        faq_id = faq_service.add_faq(
            question=question,
            answer=answer,
            category=faq_category,
            keywords=keywords,
            related_questions=related_questions,
            examples=examples,
            difficulty=difficulty
        )
        
        return {
            'id': faq_id,
            'message': 'FAQ added successfully',
            'status': 'success'
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error adding FAQ: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to add FAQ: {str(e)}")


@router.put("/{faq_id}")
async def update_faq(
    faq_id: str,
    question: Optional[str] = None,
    answer: Optional[str] = None,
    keywords: Optional[List[str]] = None,
    related_questions: Optional[List[str]] = None,
    examples: Optional[List[str]] = None,
    difficulty: Optional[str] = None
) -> Dict[str, Any]:
    """Update an existing FAQ item."""
    try:
        # Update FAQ
        success = faq_service.update_faq(
            faq_id=faq_id,
            question=question,
            answer=answer,
            keywords=keywords,
            related_questions=related_questions,
            examples=examples,
            difficulty=difficulty
        )
        
        if not success:
            raise HTTPException(status_code=404, detail=f"FAQ not found: {faq_id}")
        
        return {
            'id': faq_id,
            'message': 'FAQ updated successfully',
            'status': 'success'
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating FAQ {faq_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to update FAQ: {str(e)}")


@router.delete("/{faq_id}")
async def delete_faq(faq_id: str) -> Dict[str, Any]:
    """Delete an FAQ item."""
    try:
        # Delete FAQ
        success = faq_service.delete_faq(faq_id)
        
        if not success:
            raise HTTPException(status_code=404, detail=f"FAQ not found: {faq_id}")
        
        return {
            'id': faq_id,
            'message': 'FAQ deleted successfully',
            'status': 'success'
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting FAQ {faq_id}: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to delete FAQ: {str(e)}")


@router.get("/export/json")
async def export_faqs() -> Dict[str, Any]:
    """Export all FAQs to JSON format."""
    try:
        export_data = faq_service.export_faqs()
        return export_data
        
    except Exception as e:
        logger.error(f"Error exporting FAQs: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to export FAQs: {str(e)}")


@router.post("/import/json")
async def import_faqs(faq_data: Dict[str, Any]) -> Dict[str, Any]:
    """Import FAQs from JSON format."""
    try:
        imported_count = faq_service.import_faqs(faq_data)
        return {
            'imported_count': imported_count,
            'message': f'Successfully imported {imported_count} FAQs',
            'status': 'success'
        }
        
    except Exception as e:
        logger.error(f"Error importing FAQs: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to import FAQs: {str(e)}")


@router.get("/help/quick-start")
async def get_quick_start_guide() -> Dict[str, Any]:
    """Get quick start guide for new users."""
    try:
        guide = {
            'title': 'FinMDA-Bot Quick Start Guide',
            'steps': [
                {
                    'step': 1,
                    'title': 'Upload Your Documents',
                    'description': 'Upload PDF, Excel, or CSV files containing your financial data',
                    'example': 'Upload your income statement or balance sheet'
                },
                {
                    'step': 2,
                    'title': 'Ask Questions',
                    'description': 'Use natural language to ask questions about your financial data',
                    'example': 'What is the profit margin trend?'
                },
                {
                    'step': 3,
                    'title': 'Get Insights',
                    'description': 'Receive detailed analysis, ratios, and forecasts',
                    'example': 'Show me the financial ratios and trends'
                },
                {
                    'step': 4,
                    'title': 'Use Voice Assistant',
                    'description': 'Speak your questions using the voice assistant feature',
                    'example': 'Click the microphone and ask: "What are the key financial metrics?"'
                }
            ],
            'common_questions': [
                'What is FinMDA-Bot?',
                'How do I upload documents?',
                'What financial ratios can you calculate?',
                'Can you predict future performance?',
                'How do I use the voice assistant?'
            ],
            'tips': [
                'Upload clear, high-quality documents for better results',
                'Ask specific questions for more accurate answers',
                'Use the voice assistant for hands-free analysis',
                'Check the FAQ section for common questions',
                'Explore different financial analysis features'
            ]
        }
        
        return guide
        
    except Exception as e:
        logger.error(f"Error getting quick start guide: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to get quick start guide: {str(e)}")
